### ISSUE TITLE
The wrong aspect in games

### ISSUE THREAD

**Issue by maximus105** [joncampbell123/dosbox-x#1520]
In the Dyna Blaster game, the image is stretched horizontally (pay attention to any circle in the game) when setting fullscreen = true, output = openglnb, aspect = true
![DOSBox_2020_05_07_15_25_25_115-](https://user-images.githubusercontent.com/58270614/81298230-29833d80-907d-11ea-9eee-94dae5d3b123.png)

Аt the same settings, in the DOSBox-ECE, it looks right
![DOSBox_2020_05_07_15_25_25_115](https://user-images.githubusercontent.com/58270614/81298243-2e47f180-907d-11ea-9346-f92ec1cf8849.png)

In The Incredible Machine game, the image is compressed horizontally (only in the game screen saver, then the image is correct), (pay attention to any circle in the game) when setting fullscreen = true, output = openglnb, aspect = true.
![DOSBox_2020_05_07_15_26_05_161-](https://user-images.githubusercontent.com/58270614/81298320-4e77b080-907d-11ea-8d38-6877b7b6fa28.png)

Аt the same settings, in the DOSBox-ECE, it looks right
![DOSBox_2020_05_07_15_26_05_161](https://user-images.githubusercontent.com/58270614/81298333-520b3780-907d-11ea-9b9b-17159496cebb.png)

**Comment by rderooy** [comment]
Are you sure you copied the correct DynaBlaster screenshots above? To me they look identical (unlike TIM, which is clearly different).

**Comment by rderooy** [comment]
Are you using the latest DOSBox ECE, or the older r4301, which still had the "pixel perfect scaling" patch, which was removed in later builds?

**Comment by maximus105** [comment]
> Are you using the latest DOSBox ECE, or the older r4301, which still had the "pixel perfect scaling" patch, which was removed in later builds?

Yes. I correctly added screenshots. Pay attention to the circumference of the "bomb." She looks stretched horizontally. I tested this on the DOSBox ECE r4160 and r4338 builds. The aspect is working correctly.

**Comment by BridgeHeadland** [comment]
If I'm not mistaken, there should be a 640x7 black border to at the bottom of the The Incredible Machine picture.

**Comment by maximus105** [comment]
Yes, there is a black border in both DOSBox-X and DOSBox-ECE in TIM. But the aspect in the TIM is displayed incorrectly in DOSBox-X. Pay attention to the circumference of the ball. In DOSBox-ECE the ball looks correct and has the correct circumference. In DOSBox-X, the ball looks wrong and has a vertically elongated circle.

DOSBox ECE r4473:
![2022-04-09_103151](https://user-images.githubusercontent.com/58270614/162562003-6e347ca8-c2e7-4f78-b1e3-58a723dedf9d.png)

DOSBox-X MinGW SDL1 x64 v0.83.24:
![2022-04-09_103301](https://user-images.githubusercontent.com/58270614/162562053-5674ba11-be0a-45e4-982c-62160c953ade.png)

**Comment by BridgeHeadland** [comment]
> Ja, det er en svart kant i både DOSBox-X og DOSBox-ECE i TIM. Men aspektet i TIM vises feil i DOSBox-X. Vær oppmerksom på omkretsen av ballen. I DOSBox-ECE ser ballen riktig ut og har riktig omkrets. I DOSBox-X ser ballen feil ut og har en vertikalt langstrakt sirkel.
> 
> DOSBox ECE r4473: ![2022-04-09_103151](https://user-images.githubusercontent.com/58270614/162562003-6e347ca8-c2e7-4f78-b1e3-58a723dedf9d.png)
> 
> DOSBox-X MinGW SDL1 x64 v0.83.24: ![2022-04-09_103301](https://user-images.githubusercontent.com/58270614/162562053-5674ba11-be0a-45e4-982c-62160c953ade.png)

https://github.com/joncampbell123/dosbox-x/issues/3396

That's why I linked to an article similar to this in the thread I am linking to now.

**Comment by BridgeHeadland** [comment]
![image](https://user-images.githubusercontent.com/85258509/162615034-5faadde1-429c-437c-b29d-42ccd16ec068.png)

This DOS game image above (from The Incredible Machine) is in 471p, and since 471 is an odd number, and that this resolution is too large to inflate to the double resolution (942), then I reckon that there should be a black field of 9p, either above or below - I'm actually unsure where the field should be (I think I readed any place on the web where the black field should be at the below).

P.S. I now see that the resolution of The Incredible Machine is otherwise at 640x448 and 640x400, I was confused it with The Incredible Machine 2, where the game is at 640x441.

I reckon there should be a black field of 7p above or below the image of TIM2. I also see that the balls in the game have perfect circles in pixel perfect, than if the image is stretched to 4:3, which tends to be aspect ratio to the screens of the CRT monitors, so therefore I have reason to believe that there should be black fields at 16p, both below and above, in addition to the one black field at 7p.

I think there should be black fields both below and above Jazz Jackrabbit as well, where the playing part is 320x199. the resolution is low enough to inflate the image to double (640x398), with black bars of 41p each (so it becomes 640x480), both above and below. I think the black fields are parts of the playing part image after all.

![image](https://user-images.githubusercontent.com/85258509/162616169-40bef9b2-275a-42cf-8cbd-5509fd7a6690.png)
![image](https://user-images.githubusercontent.com/85258509/162616188-052f5dd3-7706-40eb-bc73-f55fc98407a0.png)

**Comment by maximus105** [comment]
The essence of the problem is that in DOSBox-ECE the aspect in these games is displayed correctly, but incorrectly in DOSBox-X. It seems that some patch has been applied in DOSBox-ECE to solve this problem. It would be nice to make the same patch for DOSBox-X.

**Comment by Wengier** [comment]
As with issue #3119, you can set the ``aspect_ratio`` option to a different aspect ratio (other than 4:3) with the latest build. Set ``aspect_ratio=-1:-1`` for the original image ratio. It is not the default option, but allows you to change the aspect ratio to match your games or programs better when aspect ratio correction is enabled. Hope this helps.

**Comment by BridgeHeadland** [comment]
> ![image](https://user-images.githubusercontent.com/85258509/162615034-5faadde1-429c-437c-b29d-42ccd16ec068.png)
> 
> This DOS game image above (from The Incredible Machine) is in 471p, and since 471 is an odd number, and that this resolution is too large to inflate to the double resolution (942), then I reckon that there should be a black field of 9p, either above or below - I'm actually unsure where the field should be (I think I readed any place on the web where the black field should be at the below).
> 
> P.S. I now see that the resolution of The Incredible Machine is otherwise at 640x448 and 640x400, I was confused it with The Incredible Machine 2, where the game is at 640x441.
> 
> I reckon there should be a black field of 7p above or below the image of TIM2. I also see that the balls in the game have perfect circles in pixel perfect, than if the image is stretched to 4:3, which tends to be aspect ratio to the screens of the CRT monitors, so therefore I have reason to believe that there should be black fields at 16p, both below and above, in addition to the one black field at 7p.
> 
> I think there should be black fields both below and above Jazz Jackrabbit as well, where the playing part is 320x199. the resolution is low enough to inflate the image to double (640x398), with black bars of 41p each (so it becomes 640x480), both above and below. I think the black fields are parts of the playing part image after all.
> 
> ![image](https://user-images.githubusercontent.com/85258509/162616169-40bef9b2-275a-42cf-8cbd-5509fd7a6690.png) ![image](https://user-images.githubusercontent.com/85258509/162616188-052f5dd3-7706-40eb-bc73-f55fc98407a0.png)

https://www.vogons.org/viewtopic.php?p=384596#p384596
For those of you who have not experienced it before, this page proves that the gameplay part of Jazz Jackrabbit is run in letterbox format, on an old CRT monitor... or maybe is it the hardware that makes the gameplay part run in letterbox format? I do not know, but maybe someone else, who knows more about computers and monitors, knows more about this better than me. The game probably runs in the resolutions 640x400 and 640x398 on old PCs, since the first VGA hardware is the first hardware that runs VGA games, and which, as far as I know, doubles the resolutions of the VGA games, 
unless the resolution is 400p or higher. The black fields will then correspond to 640x41 regardless. If the game runs at 320x200 and 320x199, the black bars will correspond to 320x20.5, but we know that there are no incomplete pixels at all, and I almost think that the game, with a black line of 320x1 on the top, might be stretched 20% vertically. I remember experiencing a black line at the top of the gameplay image earlier, when I played JJ in DOSBox SVN Daum, and I think I had not activated that doublescan back then.

**Comment by BridgeHeadland** [comment]
I see that the aspect ratio options, in addition to the original ratio and set ratio, which are in the latest versions of MinGW builds are based on the aspect ratio examples I mentioned, but there are several DOS-related aspect ratios, depending on what software or setup you are using. Here is a list of all the aspect ratios I know of (plus examples of screen resolutions for these aspect ratios). Now I do not know if these ratio calculations have correct names, I have calculated them from https://calculateaspectratio.com/:

1: 1 (256x256)
4: 3 (320x240, 640x480, 512x384, 800x600, 1024x768, 1600x1200)
4: 5 (320x400)
4: 6 (320x480)
6: 4 (720x480)
8: 7 (256x224)
10: 7 (320x224, 640x448)
12: 8 (720x480)
16: 5 (320x100, 640x200)
16:10 (160x100, 320x200, 640x400)
16:15 (256x240)
16:20 (160x200, 320x400)
16:24 (320x480)
18x10 (360x200, 720x400)
18:12 (360x240)
18:20 (360x400)
18:24 (360x480)
25:16 (400x256)
32:25 (256x200, 512x400)
32:29 (320x290)
32:35 (320x350)
36:35 (360x350)
39:25 (312x200)
40:23 (320x184)
64:29 (320x145)
64:35 (640x350)
72:35 (720x350)
72:61 (288x244)
160: 95 (320x190)
304: 215 (408x430)
320: 191 (320x191)
320: 199 (320x199)
384: 257 (384: 257)
640: 441 (640x441)
640: 471 (640x471)

**Comment by Wengier** [comment]
I added the 3:2 aspect ratio, which should be the same as 6:4 and 12:8 listed above. I don't think most people will ever need all the ratios, but if they do they can always set customized ratios themselves.
